### ISSUE TITLE
Yolov7-tiny  to TensorflowLite conversion results in a dynamic output model incompatible with TfLite Java API

### ISSUE THREAD

**Issue by ingura** [PINTO0309/onnx2tf#159]
### Issue Type

Others

### onnx2tf version number

1.5.36

### onnx version number

1.12.0

### tensorflow version number

2.10.1

### Download URL for ONNX

pip install onnx==1.12.0

### Parameter Replacement JSON

```json
none
```

### Description

Hi, your library is awesome! 

I converted the Yolov7-tiny from PyTorch to TfLite using: 
`onnx2tf -i yolov7-tiny.onnx -o models-NHWC-final/ -osd -oh5 -cotof`

I am trying to use it on an android device. The model works when tested on a PC however the Tensorflow Java API for Android does not support dynamic output models according to their documentation: https://www.tensorflow.org/lite/guide/inference.  While the resulting yolo tflite model has a dynamic number of outputs( the number of outputs change with the number of indications/detentions)

On the other hand, if I follow the conversion path PyTorch -> ONNX-> Tensorflow I do get yolov7 with a fixed output size so I suspect it is possible to achieve this with onnx2tf as well while also doing the NcHW to NHWc conversion in the process.

Is there a way to have onnx2tf output a fixed/static output .tflite model for yolov7-tiny?

Thank you

**Comment by PINTO0309** [comment]
I have already implemented almost all the features that engineers around the world need. Please read the README seriously first.

https://github.com/PINTO0309/onnx2tf#cli-parameter

![image](https://user-images.githubusercontent.com/33194443/215631712-ebfd36be-e418-45cb-a4b5-ec08274d99d4.png)

**Comment by ingura** [comment]
Thanks for your prompt reply. 

I went trough all the paragraphs included in the README file that mentions the "output" of the network and I could not find anything related to fixing the output size or choosing a static output size. As I understand the "-ois" parameter mentioned above overrides the input shape but not the output.

To generate the ONNX model I use the following line:
`python export.py --weights yolov7-tiny.pt --grid --end2end --simplify --topk-all 100 --iou-thres 0.65 --conf-thres 0.35 --img-size 640 640 --max-wh 640`
where "export.py" comes from the standard yolov7 distribution:  https://github.com/WongKinYiu/yolov7/blob/main/export.py

export.py generates my yolov7-tiny.onnx model. The following line takes the resulting yolov7-tiny.onnx and generates a fixed output tensorflow model which runs well on a PC (but the input is NcHW):
`onnx-tf convert -i yolov7-tiny.onnx -o model_tf/`

So the yolov7-tiny.onnx model I use with onnx2tf  should be functional. The parameters you exposed for onnx2tf are indeed many however I could not find anything related to fixing the output size of the model. 

Have i missed anything important?

**Comment by PINTO0309** [comment]
I understood. I will add a parameter to convert to fixed size NMS in the next version of the tool. Are you satisfied with just being able to specify the number of boxes to output?
![image](https://user-images.githubusercontent.com/33194443/215642034-64aa90bc-eac9-4078-8807-85fad37a3d37.png)

**Comment by ingura** [comment]
The tensor-flow API for Android basically requires a fixed size output tensor of shape [num_boxes, 7]  so by fixing the number of output boxes should solve the problem. 

Specifically the output tensor will always have the shape of [100, 7] regardless of how many objects are in the input image

**Comment by PINTO0309** [comment]
Yes, I am. I have been utilizing TensorFlow quite a bit for a long time, so I understand the specifications you gave me. 

> This is because there are more complex models that come with a lot of post-processing behind the NMS. YOLOvN post-processing is too simple.

This is a very big problem with the proposed specifications.

**Comment by ingura** [comment]
Well if all these models have a dynamic output they can not be run as a .tflite model on an Android platform. So the conversion to .tflite would not be very useful without a static output.  

Knowing that there are models that are very difficult to convert their output to be static, what if this conversion option is offered for the many models that do not have a lot of processing behind the NMS block but not the other models?  

Basically we have a parameter that allows us to fix the output size for the majority of simpler models (in regards to their NMS aspect) so they can be used as .tflite models on Androids. And to let the users know what to expect when they use this parameter a warning is displayed that lets them know that not all models can reasonably have their output fixed.  This way you have most utility embedded in your library while avoiding the problems of misuse.

**Comment by PINTO0309** [comment]
OK. then I will look into how to add the feature as a special option.

**Comment by ingura** [comment]
Thanks

**Comment by PINTO0309** [comment]
Idea notes for implementation, taking a break from my day job.
1. Delete all post-processing added on the YOLOvN side. This is because the post-processing given by the official paper implementation is quite redundant.
    - e.g. YOLOv7
    ![image](https://user-images.githubusercontent.com/33194443/215749379-6c2a61ce-0a78-4286-adcb-897d5fd5f3da.png)
2. Allow the output OP name of the model body to be specified.
3. Delete all OPs after the output OP name specified by the user.
4. Allow users to specify the maximum output bounding box, NMS threshold, number of classes, and score threshold.
5. Generate a clean post-processing of my proprietary implementation.
6. Allow users to choose which of the repertoire of post-processes to use for YOLO, SSD, and other object detection.
7. For mobile devices, or low-spec edge devices, having the number of output bounding boxes fixed at 100 is very detrimental to performance. Therefore, the post-processing is rewritten according to the maximum number of detection bounding boxes specified by the user.
8. If the final output is less than 100, zero-pad or `-1` pad all portions of `(100 - number of detections)`.
9. I am not sure if the `7` in `[num_boxes, 7]` really needs to be `7`. If possible, I do not want to output useless tensors.

**Comment by PINTO0309** [comment]
@ingura 
What exactly is `The tensorflow API for Android` that you presented, and which document specifically is the best to refer to for understanding? Actually, I would like to know the breakdown of the 7 in [num_boxes, 7] that you are expecting, rather than the API spec.
https://developer.android.com/ndk/guides/neuralnetworks

```
[X1, Y1, X2, Y2, ClassID, ClassScore, ???] <- valid_box or invalid_box?
```

**Comment by ingura** [comment]
Your implementation details are awesome.

About the API I am using Java version of tensorflow-lite-gpu. Here is my TensorFlow dependency list:
 `
    implementation "org.tensorflow:tensorflow-lite:${tflite_version}"
    implementation "org.tensorflow:tensorflow-lite-gpu:${tflite_version}"
    implementation "org.tensorflow:tensorflow-lite-gpu-api:${tflite_version}"
`
Each model can have a different output structure but in the yolov7 case our output tensor has the shape [num_boxes, 7].  The result for each box is an array of size 7 that contains:

` [batch_number, boxLeftLimitX , boxTopMostY , boxRightLimitX , boxBotomLimitY, ClassID, ClassScore]` 

Thank you for your effort!

**Comment by PINTO0309** [comment]
Thank you. I understood.

I will get around to implementation, but major concerns remain. I have a recent history of supporting implementations of models that can use Android GPUs, but it did not work. For more information, please see the issue below.

[I cannot run a transformer model with token-level output on accelerated hardware in TF Lite #59232](https://github.com/tensorflow/tensorflow/issues/59232)

Currently, GPU Delegate only supports the following very limited basic OPs. Also, `Gather` cannot be used and there are significant restrictions on the use of `strided_slice`. Therefore, the post-processing to be implemented based on this idea will be a model with fallback to the CPU or GPU Delegate may generate an error and Abort.

https://www.tensorflow.org/lite/performance/gpu

```
ADD
AVERAGE_POOL_2D
CONCATENATION
CONV_2D
DEPTHWISE_CONV_2D v1-2
EXP
FULLY_CONNECTED
LOGISTIC
LSTM v2 (Basic LSTM only)
MAX_POOL_2D
MAXIMUM
MINIMUM
MUL
PAD
PRELU
RELU
RELU6
RESHAPE
RESIZE_BILINEAR v1-3
SOFTMAX
STRIDED_SLICE
SUB
TRANSPOSE_CONV
```
Just my guess before doing the implementation, but I am thinking that eventually you may need to re-implement the post-processing part with a custom GPU Delegate or custom operation.

**Comment by ingura** [comment]
If the GPU support is not there let's have it running on the CPU and see if it is comparable with yolov4 in terms of computational performance. And yolov4 did pretty well. 

The paper mentions that there is a significant computational reduction happening in yolov7 compared to v4

**Comment by ingura** [comment]
Not always efficient but tfLite can split the inference execution between the GPU and CPU: https://www.tensorflow.org/lite/performance/gpu

If some of the ops are not supported by the GPU delegate, the framework will only run a part of the graph on the GPU and the remaining part on the CPU.

**Comment by PINTO0309** [comment]
I see. I'll try to implement it before worrying about this and that.

I don't have an environment for testing on Android, so I would be happy to have you help me test the tool once I have finished my experimental customization of the tool.

**Comment by ingura** [comment]
I look forward to it

**Comment by PINTO0309** [comment]
![image](https://user-images.githubusercontent.com/33194443/216043312-65d45eb7-76ac-4d77-9829-a1f04ff367fb.png)

**Comment by ingura** [comment]
Ill give it a try with GPU support and without.

**Comment by PINTO0309** [comment]
I am trying to apply partial optimization first and then improve the whole thing at once. Several more pull requests will be issued, but please wait a little longer.

**Comment by PINTO0309** [comment]
ONNX exported with `--end2end`. Just fixed the output of YOLOv7 to `[100, 7]`. I do not expect it to work properly. 
Experimental: `[100, 7]`
[yolov7-tiny_float16.tflite.zip](https://github.com/PINTO0309/onnx2tf/files/10569726/yolov7-tiny_float16.tflite.zip)
[yolov7-tiny_float32.tflite.zip](https://github.com/PINTO0309/onnx2tf/files/10569712/yolov7-tiny_float32.tflite.zip)
![image](https://user-images.githubusercontent.com/33194443/216345181-b218b89f-1d5c-48c1-a08e-8913e54b9ab4.png)

**Comment by PINTO0309** [comment]
In fact, that is the last model I posted that implements it. I see that it is an error... :thinking: 
```python
        # YOLOv7 Special fixed outputs
        max_output_boxes_per_class = 100
        final_output = outputs[0]
        output_paddings = tf.zeros(shape=[max_output_boxes_per_class, 7], dtype=tf.float32)

        indices = tf.range(0, tf_shape(input_tensor=final_output)[0])
        outputs = [
            tf.tensor_scatter_nd_update(
                tensor=output_paddings,
                indices=indices,
                updates=final_output,
            )
        ]
```
![image](https://user-images.githubusercontent.com/33194443/216471610-61c93e65-0e28-4e46-92bb-88e0b01191a0.png)

**Comment by Hyunseok-Kim0** [comment]
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/delegates/gpu/gl/kernels

Tensorflow GPU delegates supports only limited operations. I guess it is almost impossible to implement NMS using those. It looks even `tf.range` is not supported, which makes almost impossible to implement neccesary `top-k` or `sort` action for NMS.

![Screenshot from 2023-02-02 17-26-34](https://user-images.githubusercontent.com/34959032/216471783-fd2eee8c-5169-4b2c-baa1-c1d67ca16e8c.png)

There are two options for now.
1. If final goal is only replacing `NonMaxSuppression` to other operations, it is possible by using `tf.image.non_max_suppression_padded`. 
The implementation using several sub-operations unlike `non_max_suppression_v4` 
 as shown above.

<img src="https://user-images.githubusercontent.com/34959032/216472687-ce166107-8196-4a0a-b2b2-b50eed032698.png" height="350">

2. If you want only static output shape, `non_max_suppression_v4` has `pad_to_max_output_size` option.
For now, `NonMaxSuppression.py` passing `False` and using slice to remove excess indices. After making option for static output for NMS, it is possible to make user to determine maximum box number.

**Comment by PINTO0309** [comment]
Thanks @Hyunseok-Kim0.
I'll give it a try.

**Comment by PINTO0309** [comment]
Excellent.
- `pad_to_max_output_size = False`
  ![image](https://user-images.githubusercontent.com/33194443/216474974-42f488ba-b43c-4cd8-af3f-abca31b44550.png)
- `pad_to_max_output_size = True`
  ![image](https://user-images.githubusercontent.com/33194443/216475063-58679c2e-5cc1-409b-a895-4ecc6f5ca771.png)

**Comment by Hyunseok-Kim0** [comment]
You have to disable slice when using that option, still the dynamic output is generated in the image because of `num_valid` 

```
if pad_to_max_output_size:
    return selected_indices

else:
    return selected_indices[:num_valid]
```

**Comment by PINTO0309** [comment]
OK. Thanks.
![image](https://user-images.githubusercontent.com/33194443/216476007-d4a8d582-3639-4043-b687-8649e7a982e4.png)

**Comment by ingura** [comment]
> When the output of `NonMaxSuppressionV4` is variable, the runtime seems to output an error when padding because of the variable number of outputs, which can be zero or 100 depending on the input image. It seems necessary to give up using `NonMaxSuppressionV4`.

It seems like the conversion to a static output of a dynamic output network by padding its dynamic output is quite a general approach and it is worth looking into the cause of this error.   

If the error happens at the padding step could it be because it cant handle padding with a meaningless 0 size array? Or maybe the issue happens at the other extreme pad(100+)?

**Comment by PINTO0309** [comment]
Maybe my use of tf.sensor_scatter_nd_update is just wrong.

In any case, your most recent objective of generating a fixed size output is achieved by merging my pull request above.

We have completed the first step in order to address the issues in a detailed step-by-step process. I will look into the issue of padding errors due to self-preprocessing later.

**Comment by ingura** [comment]
Thanks!

**Comment by PINTO0309** [comment]
I made a mistake in releasing the package, so the PyPI version and the Docker version mismatched, but I just released the latest version 1.5.39.
https://github.com/PINTO0309/onnx2tf/releases/tag/1.5.39

1.5.38 is synonymous with 1.5.39 for Docker.

[yolov7-tiny_org_post_float16_fixed.tflite.zip](https://github.com/PINTO0309/onnx2tf/files/10575362/yolov7-tiny_org_post_float16_fixed.tflite.zip)
[yolov7-tiny_org_post_float32_fixed.tflite.zip](https://github.com/PINTO0309/onnx2tf/files/10575363/yolov7-tiny_org_post_float32_fixed.tflite.zip)
![image](https://user-images.githubusercontent.com/33194443/216508134-29eff3a4-c2b1-47e0-906b-e733033ba534.png)

**Comment by ingura** [comment]
You're amazing, you fixed it! Maybe in the future the GPU Delegate will be better supported as well. 

Thank you very much!
